staging/src/k8s.io/component-base/metrics/metric.go

```diff
@@ -210,6 +211,31 @@ func (c *selfCollector) Collect(ch chan<- prometheus.Metric) {
 	ch <- c.metric
 }
 
+// metricWithExemplar is an interface that knows how to attach an exemplar to certain supported metric types.
+type metricWithExemplar interface {
+	withExemplar(v float64)
+}
+
+// exemplarMetric is a holds a context to extract exemplar labels from, and a metric to attach them to. It implements the metricWithExemplar interface.
+type exemplarMetric struct {
+	*Counter
+}
+
+// withExemplar attaches an exemplar to the metric.
+func (e *exemplarMetric) withExemplar(v float64) {
+	var exemplarLabels prometheus.Labels
+	maybeSpanCtx := trace.SpanContextFromContext(e.ctx)
+	if maybeSpanCtx.IsValid() {
+		exemplarLabels = prometheus.Labels{
+			"trace_id": maybeSpanCtx.TraceID().String(),
+			"span_id":  maybeSpanCtx.SpanID().String(),
+		}
+	}
+	if m, ok := e.CounterMetric.(prometheus.ExemplarAdder); ok {
```

Comment:
Is the `CounterMetric` always going to be an `ExemplarAdder`?  Otherwise, it would be good to skip the creation of the exemplarLabels.